app/src/main/java/org/sil/gatherwords/SessionActivity.java

```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
         startActivity(i);
     }
 
-    private static class InsertSessionsTask extends AsyncTask<Session, Void, Void> {
+    private static class InsertSessionsTask extends AsyncTask<Session, Void, List<Long>> {
         private SessionDao sDAO;
 
         InsertSessionsTask(AppDatabase db) {
             sDAO = db.sessionDao();
         }
 
         @Override
-        protected Void doInBackground(Session... sessions) {
-            sDAO.insertSession(sessions);
+        protected List<Long> doInBackground(Session... sessions) {
+            return sDAO.insertSession(sessions);
+        }
+
+    }
+
+    /**
+     * Async task for inserting words from the selected word list
+     */
+    private static class InsertWordsTask extends AsyncTask<Void, Double, List<Long>> {
+        private WordDao wordDao;
+        private AsyncTask<Session, Void, List<Long>> task;
+        private final ThreadLocal<SessionActivity> sessionActivity = new ThreadLocal<>();
+        private Long sessionID;
+        private int maxProgress;
+
+        InsertWordsTask(AsyncTask<Session, Void, List<Long>> sessionTask, SessionActivity activity) {
+            sessionActivity.set(activity);
+            wordDao = sessionActivity.get().db.wordDao();
+            task = sessionTask;
+        }
+
+        @Override
+        protected List<Long> doInBackground(Void... voids) {
+            List<Long> ids = new ArrayList<>();
+            try {
+                // Wait for the insert session task to finish and get the result (the id(s) of the session)
+                ids = task.get();
+            } catch (InterruptedException | ExecutionException e) {
+                Log.getStackTraceString(e);
+            }
+            if (ids.size() == 1) {
+                sessionID = ids.get(0);
+            }
+            insertFromAsset();
+
             return null;
         }
+
+        @Override
+        protected void onProgressUpdate(Double... values) {
+            super.onProgressUpdate(values);
+            double progress = (values[0] / maxProgress)*100;
+            CharSequence sequence = new DecimalFormat("##.##%").format(progress);
+            // Too slow!!! TODO get quick progress updates
+//            Toast.makeText(sessionActivity.getApplicationContext(), sequence, Toast.LENGTH_SHORT).show();
+        }
+
+        /**
+         * Insert the words from the selected asset or exit if none selected
+         */
+        private void insertFromAsset() {
+            if (sessionActivity.get().worldListToLoad.equals("")) {
```

Comment:
Just `.get()` it once per function and store it in a variable.

Reply:
How about getting it only when I need the context (which has the data leak), and the other variables are stored on construction